sa/database.go

```diff
@@ -156,12 +160,103 @@ func NewDbMapFromConfig(config *mysql.Config, settings DbSettings) (*boulderDB.W
 	dbmap := &gorp.DbMap{Db: db, Dialect: dialect, TypeConverter: BoulderTypeConverter{}}
 
 	initTables(dbmap)
-
 	return &boulderDB.WrappedMap{DbMap: dbmap}, nil
 }
 
-// adjustMySQLConfig sets certain flags that we want on every connection.
-func adjustMySQLConfig(conf *mysql.Config) {
+var checkQuoteRE = regexp.MustCompile(`^'{1}\w+'{1}$`)
+var checkImproperIntRE = regexp.MustCompile(`^'{1}\d+'{1}$`)
+var checkFloatOrIntRE = regexp.MustCompile(`^?\d+(\.\d+)?$`)
+
+// checkMariaDBSystemVariableQuoting ensures that a MariaDB config passed in via
+// SA setDefault or via DSN have the correct quoting for string variables,
+// enumerations as strings, and enumerations as integers/floating points to
+// prevent future problems such as PR #6683 from occurring. This is a manually
+// curated list of server variables generated by a tool in issue #6687.
+func checkMariaDBSystemVariableQuoting(name string, value string) error {
+	mariaDBStringTypes := []string{
+		"bind_address",
+		"character_set_client",
+		"character_set_connection",
+		"character_set_database",
+		"character_set_filesystem",
+		"character_set_results",
+		"character_set_server",
+		"character_set_system",
+		"collation_connection",
+		"collation_database",
+		"collation_server",
+		"debug_sync",
+		"enforce_storage_engine",
+		"external_user",
+		"ft_boolean_syntax",
+		"hostname",
+		"ignore_db_dirs",
+		"init_connect",
+		"lc_time_names",
+		"license",
+		"old_mode",
+		"optimizer_switch",
+		"proxy_user",
+		"session_track_system_variables",
+		"shared_memory_base_name",
+		"sql_mode",
+		"system_time_zone",
+		"time_zone",
+		"version",
+		"version_comment",
+		"version_compile_machine",
+		"version_compile_os",
+		"version_malloc_library",
+		"version_source_revision",
+	}
+
+	mariaDBEnumTypes := []string{
+		"completion_type",
+		"concurrent_insert",
+		"default_storage_engine",
+		"default_tmp_storage_engine",
+		"delay_key_write",
+		"event_scheduler",
+		"optimizer_trace",
+		"plugin_maturity",
+		"session_track_transaction_info",
+		"tx_isolation",
+	}
+
+	for _, v := range mariaDBStringTypes {
+		if name == v {
+			if checkQuoteRE.FindString(value) == value {
+				return nil
+			}
+			return fmt.Errorf("%v=%v string is not properly quoted", name, value)
+		}
+	}
+
+	for _, v := range mariaDBEnumTypes {
+		if name == v {
+			if checkFloatOrIntRE.FindString(value) == value {
+				return nil
+			} else if checkImproperIntRE.FindString(value) == value {
+				return fmt.Errorf("%v=%v enumeration as integer is quoted, but should not be", name, value)
+			} else if checkQuoteRE.FindString(value) == value {
+				return nil
+			}
+			return fmt.Errorf("%v=%v enumeration as string is not properly quoted", name, value)
```

Comment:
This sequence of conditionals is not obvious to the reader, and has left me somewhat confused. I totally believe that the logic is correct, but it could benefit from some comments addressing what the regex is checking at each step, and why that check is correct/important. For example:
- The first check is "it's okay if this enum value is an unquoted int or float". Are floats acceptable enum values? That's surprising to me.
- The second check doesn't require that floats be unquoted. Should it?

Is there a way to simplify these, by separating the list of "enum" names into "int-valued-enum" and "string-valued-enum", and then running their prospective values through just one regex each?

Reply:
I agree with the separation being proposed here.